crate/kmip/src/crypto/rsa/ckm_rsa_pkcs.rs

```diff
@@ -0,0 +1,153 @@
+//! Implements the RSA Key Encryption Mechanism `CKM_RSA_PKCS`
+//! a.k.a PKCS #1 RSA V1.5 as specified in PKCS#11 v2.40 available at
+//! https://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/cos01/pkcs11-curr-v2.40-cos01.html#_Toc408226893
+//!
+//! This scheme is no longer FIPS approved for wrap/unwrap encrypt/decrypt operations.
+use openssl::{
+    pkey::{PKey, Private, Public},
+    pkey_ctx::PkeyCtx,
+};
+use zeroize::Zeroizing;
+
+#[cfg(feature = "fips")]
+use super::FIPS_MIN_RSA_MODULUS_LENGTH;
+use crate::error::KmipError;
+#[cfg(feature = "fips")]
+use crate::kmip_bail;
+
+/// Key Wrap using `CKM_RSA_PKCS`
+/// a.k.a PKCS #1 RSA V1.5 as specified in PKCS#11 v2.40 available at
+/// https://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/cos01/pkcs11-curr-v2.40-cos01.html#_Toc408226893
+///
+/// The maximum dek length is k-11 where k is the length in octets of the RSA modulus
+/// The output length is the same as the modulus length.
+///
+/// Arguments:
+/// - `pubkey`: the public key used to wrap the key
+/// - `key_to_wrap`: the data encryption key to wrap
+pub fn ckm_rsa_pkcs_key_wrap(
+    pub_key: &PKey<Public>,
+    key_to_wrap: &[u8],
+) -> Result<Vec<u8>, KmipError> {
+    let (mut ctx, mut ciphertext) = init_ckm_rsa_pkcs_encryption_context(pub_key)?;
+    ctx.encrypt_to_vec(key_to_wrap, &mut ciphertext)?;
+    Ok(ciphertext)
+}
+
+/// Encryption using `CKM_RSA_PKCS`
+/// a.k.a PKCS #1 RSA V1.5 as specified in PKCS#11 v2.40 available at
+/// https://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/cos01/pkcs11-curr-v2.40-cos01.html#_Toc408226893
+///
+/// The maximum plaintext length is  k-11 where k is the length in octets of the RSA modulus
+/// The output length is the same as the modulus length.
+///
+/// Arguments:
+/// - `pubkey`: the public key used to wrap the key
+/// - `plaintext`: the plaintext to encrypt
+pub fn ckm_rsa_pkcs_encrypt(
+    pub_key: &PKey<Public>,
+    plaintext: &[u8],
+) -> Result<Vec<u8>, KmipError> {
+    let (mut ctx, mut ciphertext) = init_ckm_rsa_pkcs_encryption_context(pub_key)?;
+    ctx.encrypt_to_vec(plaintext, &mut ciphertext)?;
+    Ok(ciphertext)
+}
+
+fn init_ckm_rsa_pkcs_encryption_context(
+    pub_key: &PKey<Public>,
+) -> Result<(PkeyCtx<Public>, Vec<u8>), KmipError> {
+    let rsa_pub_key = pub_key.rsa()?;
+
+    // The ciphertext has the same length as the modulus.
+    let encapsulation_bytes_len = rsa_pub_key.size() as usize;
+    let ciphertext = Vec::with_capacity(encapsulation_bytes_len);
+
```

Comment:
resize() reallocates the difference, so we could avoid the pre-allocation but it does not make a difference performance wise

Reply:
Yes, no performance gain, but it would be cleaner not to pre-allocate in this function. Its name (`init_ckm_rsa_pkcs_encryption_context`) is not related to the ciphertext anyway. It would also save these 3 lines of code.

Reply:
It is not the same. Although it is not used here. If you were to do batch decryptions, you would call the `init`only once and the `encrypt_to_vec()` multiple times. The main allocation is done only once.